### PR TITLE
Add auto upgrade functionality

### PR DESCRIPTION
**Pull request by vicky-g**
This PR is to:
- Uncomment out the audius-cli ability to run auto-upgrades
- Pipe any logs to `{manifests_path}/.logs/auto_upgrade.txt`

Tests:
- Ran `audius-cli autoupgrade "* * * * *"`
- Kept track of the logs to observe behavior at every minute mark

With updates in the latest manifest repo:
```
Mon Apr 12 18:15:01 UTC 2021
Auto-merging audius-cli
Merge made by the 'recursive' strategy.
 README.md                                          | 82 +++++++++++++++-----
 audius-cli                                         | 88 +++++++++++++++++++---
 .../creator-node/creator-node-deploy-backend.yaml  |  2 +-
 .../discovery-provider-deploy-no-workers.yaml      |  4 +-
 .../discovery-provider-deploy.yaml                 |  8 +-
 audius/identity/identity-cm.yaml                   | 31 ++++++++
 audius/identity/identity-deploy.yaml               | 83 ++++++++++++++++++++
 audius/identity/identity-svc.yaml                  | 35 +++++++++
 backup.py                                          |  2 +-
 sp-utilities/claim/claim.js                        |  8 +-
 sp-utilities/discovery-provider/healthChecks.js    | 30 +++++++-
 sp-utilities/package-lock.json                     |  6 +-
 sp-utilities/package.json                          |  2 +-
 13 files changed, 332 insertions(+), 49 deletions(-)
 create mode 100644 audius/identity/identity-cm.yaml
 create mode 100644 audius/identity/identity-deploy.yaml
 create mode 100644 audius/identity/identity-svc.yaml
creator-node-backend-cm has unset keys: discoveryProviderWhitelist
Reverting audius/creator-node/creator-node-cm.yaml
Reverting audius/discovery-provider/discovery-provider-cm.yaml
Pulling changes from source
Updating config maps
Upgrading creator-node...
Do you want to continue? [y/N] service/creator-node-backend-svc unchanged
service/creator-node-cache-svc unchanged
service/creator-node-db-svc unchanged
service/creator-node-ipfs-svc unchanged
persistentvolume/creator-node-backend-pv unchanged
persistentvolumeclaim/creator-node-backend-pvc unchanged
persistentvolume/creator-node-db-pv unchanged
persistentvolumeclaim/creator-node-db-pvc unchanged
persistentvolume/creator-node-ipfs-pv unchanged
persistentvolumeclaim/creator-node-ipfs-pvc unchanged
configmap/creator-node-ipfs-cm unchanged
deployment.apps/creator-node-ipfs unchanged
configmap/creator-node-backend-cm unchanged
configmap/creator-node-db-cm unchanged
deployment.apps/creator-node-backend configured
deployment.apps/creator-node-cache unchanged
deployment.apps/creator-node-db unchanged
```

Without updates in the latest manifest repo:
```
Mon Apr 12 18:16:01 UTC 2021
Already up-to-date.
creator-node-backend-cm has unset keys: discoveryProviderWhitelist
Reverting audius/creator-node/creator-node-cm.yaml
Reverting audius/discovery-provider/discovery-provider-cm.yaml
Reverting audius/identity/identity-cm.yaml
Pulling changes from source
Updating config maps
Upgrading creator-node...
Do you want to continue? [y/N] service/creator-node-backend-svc unchanged
service/creator-node-cache-svc unchanged
service/creator-node-db-svc unchanged
service/creator-node-ipfs-svc unchanged
persistentvolume/creator-node-backend-pv unchanged
persistentvolumeclaim/creator-node-backend-pvc unchanged
persistentvolume/creator-node-db-pv unchanged
persistentvolumeclaim/creator-node-db-pvc unchanged
persistentvolume/creator-node-ipfs-pv unchanged
persistentvolumeclaim/creator-node-ipfs-pvc unchanged
configmap/creator-node-ipfs-cm unchanged
deployment.apps/creator-node-ipfs unchanged
configmap/creator-node-backend-cm unchanged
configmap/creator-node-db-cm unchanged
deployment.apps/creator-node-backend unchanged
deployment.apps/creator-node-cache unchanged
deployment.apps/creator-node-db unchanged
```

### CODE REVIEW

File: audius-cli [hunk 1]
```diff
@@ -150,8 +150,9 @@ def auto_upgrade(cron_expression, remove):
         for job in cron.find_comment("audius-cli auto-upgrade"):
             cron.remove(job)
         if not remove:
+            # Pipe auto upgrade logs to auto_upgrade_logs.txt
             job = cron.new(
-                "bash -c 'yes | audius-cli upgrade'",
+                "date >> auto_upgrade_logs.txt; (bash -c 'yes | /usr/local/bin/audius-cli upgrade' 2>&1) >> auto_upgrade_logs.txt; echo >> auto_upgrade_logs.txt",
```
Comment (vicky-g) [inline audius-cli:155]: this way, we can keep track of what is happening via this cron job + if any issues  come up, we can debug

**Review by cheran-senthil — APPROVED** [review]
Amazing! I'm very excited for this to get merged; just one last bit of testing would be to check if the auto-upgrade works when there are changes too.

However, seeing that it pulls and is already applying the kube configs afterward, I think this is ready to merge.

🚀 🚀 🚀

**Comment by vicky-g** [comment]
> Amazing! I'm very excited for this to get merged; just one last bit of testing would be to check if the auto-upgrade works when there are changes too.
> 
> However, seeing that it pulls and is already applying the kube configs afterward, I think this is ready to merge.
> 
> 🚀 🚀 🚀

in my logs, you should see that when there were changes, the command would auto pull the changes, and redeploy with the changes 😃

**Comment by cheran-senthil** [comment]
Sorry, I misinterpreted that as the first run, great work then, pretty sure this is good to ship!

Think we should put this on our nodes today/monday and observe for a while

File: audius-cli (hunk 1 above)
Comment (cheran-senthil) [reply audius-cli:155]: Maybe logging should be optional, perhaps allow path specification.

Comment (vicky-g) [reply audius-cli]: discussed offline -- should be ok to keep as the file should not get too big using the default cadence of upgrading every 6 hours and writes of 1-2kB per pull. (quick math = about 57 weeks for log file to get to 1MB)

Comment (vicky-g) [reply audius-cli]: updated the path to `{manifests_path}/.logs/auto_upgrade.txt` + added `.logs/` to .gitignore

**Review by cheran-senthil — APPROVED** [review]
Looks great, I think this is ready to deploy on stage! 🚢 

cc: @dmanjunath

**Comment by vicky-g** [comment]
In case SPs ever auto-upgrade to a failing release,  it might be useful to also add a rollback option into audius-cli. Or go with a branching solution in audius-protocol.

What do you guys think? @cheran-senthil @dmanjunath

**Comment by cheran-senthil** [comment]
I think having a develop branch would be the best way forward, the only case where an upgrade can seriously impact the network is when the upgrade changes audius-cli thereby preventing things from getting rolled back.

**Comment by cheran-senthil** [comment]
Merged because this has no diff

**Pull request merged**